### PR TITLE
Bring extents in line with R17

### PR DESCRIPTION
**Pull request by crtrott**
The biggest changes are removal of defaulted constructors/assignment
and adding from span constructor.

### CODE REVIEW

File: include/experimental/__p0009_bits/extents.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]: Is the point of testing `MDSPAN_HAS_CXX_20` just to work around an NVCC bug?  If so, then would you consider just replacing this with `#ifdef __NVCC__` for now, until we can do better testing for specifically supported NVCC versions?  If not, then would you consider using specifically relevant feature test macros, like `__cpp_concepts`?

File: include/experimental/__p0009_bits/extents.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]:
```suggestion
  extents(std::span<SizeType, N> exts) noexcept
```
The current version of the proposal in the main branch passes `span` by value, not const reference.

File: include/experimental/__p0009_bits/extents.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp:423]: Should we consider using `is_constant_evaluated()` here?
```c++
if (! std::is_constant_evaluated()) {
  assert(n < rank());
}
```
See e.g., https://godbolt.org/z/hYKT5c9qM
```c++
#include <cassert>
#include <iostream>
#include <random>
#include <type_traits>

int random_number()
{
    // Seed with a real random value, if available
    std::random_device r;
 
    // Choose a random mean between 1 and 6
    std::default_random_engine e1(r());
    std::uniform_int_distribution<int> uniform_dist(1, 6);
    int mean = uniform_dist(e1);
    return mean;
}

constexpr int foo(int k) {
    if (std::is_constant_evaluated()) {
        return -1;
    } else {
        assert(k + 1 > 0);
        return k - 2;
    }
}

template<int k = foo(50)>
struct Thing {
  Thing() { std::cout << "Thing: k = " << k << std::endl; }
};

int main() {
  const int i = random_number();
  std::cout << "foo(" << i << ") = " << foo(i) << std::endl;
  std::cout << "foo(" << 42 << ") = " << foo(42) << std::endl;
  Thing<> thing_one;
  return 0;
}
```

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp]: Would you consider using the feature test macro `__cpp_lib_span` instead?

File: tests/test_extents.cpp
Comment (mhoemmen) [inline tests/test_extents.cpp]:
```suggestion
#ifdef __cpp_lib_span
```

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp]:
```suggestion
#ifdef __cpp_lib_span
```

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp]:
```suggestion
#ifdef __cpp_lib_span
```

File: include/experimental/__p0009_bits/standard_layout_static_array.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/standard_layout_static_array.hpp]:
```suggestion
#ifdef __cpp_lib_span
```

**Review by mhoemmen — DISMISSED** [review]
Approving so you can merge right away, but there is one point of divergence from the proposal: the `extents` constructor that takes a `span` should take it by value, not const reference.

File: include/experimental/__p0009_bits/extents.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: span is C++20 ... so the constructor doesn't work if we don't have that :-)

File: include/experimental/__p0009_bits/extents.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp]: yeah will change

File: include/experimental/__p0009_bits/extents.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/extents.hpp:423]: yeah we should try that. Just need to come up with a way in the constructors to actually spell the whole condition. But we can do that in another PR later.

**Review by mhoemmen — APPROVED** [review]

**Pull request merged**
